### PR TITLE
feat: support configurable web terminal rendering

### LINKED ISSUE #10082: web terminal characters are not rendering in v2.1.5+
multiple customers have reported characters not being rendered in the web terminal until `Enter` is clicked.

we've tracked this down to a regression in the `webgl` package, which can be found here: https://github.com/xtermjs/xterm.js/issues/4665. `webgl` was introduced in Coder version 2.1.4 https://github.com/coder/coder/pull/9320. to fix this, we should default back to the previous rendering binary `canvas`, and create a deployment-wide flag to opt-in to `webgl`. the `webgl` regression should be fixed in upcoming versions of Chrome.

image

image

### PR DESCRIPTION
**Pull request by sreya**
- Added a deployment option for configuring web terminal rendering. Valid values are 'webgl', 'canvas', and 'dom'.

fixes #10082

### CODE REVIEW

File: cli/testdata/coder_server_--help.golden [hunk 1]
```diff
@@ -66,6 +66,10 @@ Clients include the coder cli, vs code extension, and the web UI.
       --ssh-hostname-prefix string, $CODER_SSH_HOSTNAME_PREFIX (default: coder.)
           The SSH deployment prefix is used in the Host of the ssh config.
 
+      --web-terminal-renderer string, $CODER_WEB_TERMINAL_RENDERER (default: webgl)
+          The framework to use when rendering the terminal. Valid values are
+          'canvas', 'webgl', or 'dom'.
+
```
Comment (bpmct) [inline cli/testdata/coder_server_--help.golden:72]: Is the plan to make canvas the default in another PR if the web terminal renderer is, in fact, the problem? I think in our case it's best to sacrifice performance in favor of reliability.

Unless we know which _specific niche cases_ lead to the rendering bug which excludes 95% of users.

Comment (ericpaulsen) [reply cli/testdata/coder_server_--help.golden:72]: i agree - we should default to `canvas` until the upstream Chromium version (119) is released that contains the fix.

Comment (sreya) [reply cli/testdata/coder_server_--help.golden:72]: I know that according to @BrunoQuaresma there are issues with both `canvas` and `webgl`. I made `webgl` the default since it's the default now but if there's an option that has the best experience I would prefer making that the default.

File: site/src/pages/TerminalPage/TerminalPage.tsx [hunk 2]
```diff
@@ -166,9 +167,13 @@ const TerminalPage: FC<TerminalPageProps> = ({ renderer }) => {
         background: colors.gray[16],
       },
     });
-    // DOM is the default renderer.
-    if (renderer === "webgl") {
+    if (config.data && config.data.config.web_terminal_renderer === "webgl") {
```
Comment (code-asher) [inline site/src/pages/TerminalPage/TerminalPage.tsx:170]: Not important or anything (if we really care we should use a linting rule), but just a note that we seem to generally prefer `?` in this situation:

```suggestion
    if (config.data?.config.web_terminal_renderer === "webgl") {
```

File: site/src/pages/TerminalPage/TerminalPage.tsx [hunk 3]
```diff
@@ -166,9 +167,13 @@ const TerminalPage: FC<TerminalPageProps> = ({ renderer }) => {
...
       terminal.loadAddon(new WebglAddon());
+    } else if (
+      config.data &&
+      config.data.config.web_terminal_renderer === "canvas"
```
Comment (code-asher) [inline site/src/pages/TerminalPage/TerminalPage.tsx:174]:
```suggestion
      config.data?.config.web_terminal_renderer === "canvas"
```

File: site/src/pages/TerminalPage/TerminalPage.tsx [hunk 4]
```diff
@@ -28,18 +29,16 @@ import {
...
   workspaceAgentErrorMessagePrefix: "Unable to fetch workspace agent: ",
   websocketErrorMessagePrefix: "WebSocket failed: ",
 };
 
-type TerminalPageProps = React.PropsWithChildren<{
-  renderer: "webgl" | "dom";
-}>;
-
-const TerminalPage: FC<TerminalPageProps> = ({ renderer }) => {
+const TerminalPage: FC = () => {
```
Comment (code-asher) [inline site/src/pages/TerminalPage/TerminalPage.tsx:41]: I think we are getting a test failure because the test looks for HTML elements so it relies on the `dom` renderer being used (see `webTerminal.spec.ts`).  Maybe we can start the e2e tests with `dom` set as the renderer?  Looks like we would do that in `playwright.config.ts` if I understand correctly.

**Review by code-asher — APPROVED** [review]
I tried it out and it works perfectly.  I think all we have to do is tweak the e2e test startup flags to use the `dom` renderer.
